PSOC6: initial integration of Cypress HAL

vmedcy opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
This PR brings the following major updates to the Cypress PSoC 6 targets for Mbed OS 5.13:

* Initial implementation of the Cypress PSoC 6 hardware abstraction layer (targets/TARGET_Cypress/TARGET_PSOC6/psoc6csp/hal)
* Ported Mbed OS HAL APIs implementation on top of the Cypress HAL (targets/TARGET_Cypress/TARGET_PSOC6/cy_*_api.c)
* Initial implementation of the Cypress Wi-Fi Host Driver (targets/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD)
* WHD implementation of EMAC driver and NetworkInterfaceAPI (features/netsocket/emac-drivers/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD)
* USB device HAL implementation
* Simplification of the CM0+ prebuilt application merging flow (replaced Python post-binary hex_filename merging hooks with prebuilt C arrays managed by the standard toolchain linker script).

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->
@ARMmbed/team-cypress 

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

vmedcy committed:
PSOC6: format extra labels for Cypress kits

vmedcy committed:
PSOC6: add psoc6cm0p asset with prebuilt CM0+ C arrays

vmedcy committed:
PSOC6: use prebuilt CM0+ C arrays instead of HEX images

Replace the prebuilt CM0+ HEX images with C files that provide the
variable cy_m0p_image placed to the combined ELF image with linker
script. This simplifies build flow of PSoC 6 application and improves
compatibility with IDE export targets. It is still possible to use
the custom prebuilt HEX images for PSA targets that remove CM0P_SLEEP
label and specify the `hex_filename` in targets.json.
Linker scripts are compatible with both scenarios.

vmedcy committed:
PSOC6: add psoc6csp asset with Cypress HAL implementation

PSoC 6 Chip Support Package provides hardware abstraction layer
for Cypress PSoC 6 device peripherals.

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/interface/WhdSTAInterface.h:
```diff
@@ -0,0 +1,117 @@
+/* WHD implementation of NetworkInterfaceAPI
+ * Copyright (c) 2017 ARM Limited
```
Please add SPDX identifier to new files (also the year 2017 is correct here?)

vmedcy replied:
Fixed by https://github.com/ARMmbed/mbed-os/pull/10692/commits/c115f23028c674329e859c26f4b7ad06f5d3d66c#diff-8d3ee587e2e526b518f9d8b2995fd36cR3

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/interface/WhdSTAInterface.h:
```diff
@@ -0,0 +1,117 @@
+#ifndef WHD_STA_INTERFACE_H
+#define WHD_STA_INTERFACE_H
+
+#include "mbed.h"
```
driver should not fetch mbed.h file but instead pull what is needed

vmedcy replied:
Fixed by https://github.com/ARMmbed/mbed-os/pull/10692/commits/c115f23028c674329e859c26f4b7ad06f5d3d66c#diff-8d3ee587e2e526b518f9d8b2995fd36cL20

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/interface/whd_emac.cpp:
```diff
@@ -0,0 +1,205 @@
+
+bool WHD_EMAC::power_up()
+{
+	drvp = *(cybsp_get_wifi_driver());
```
the code looks unaligned in this function (these 2 lines)

vmedcy replied:
Fixed by applying astyle in https://github.com/ARMmbed/mbed-os/pull/10692/commits/c115f23028c674329e859c26f4b7ad06f5d3d66c#diff-38f33b303ec6a94c6f64f1bd8778ca90R91

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/interface/whd_emac.cpp:
```diff
@@ -0,0 +1,205 @@
+{
+void cy_network_process_ethernet_data(whd_interface_t ifp, whd_buffer_t buffer)
+{
+	emac_mem_buf_t *mem_buf = NULL;
```
alignment issue here - pls run astyle with our coding style file (.astylerc)

vmedcy replied:
Fixed by applying astyle in https://github.com/ARMmbed/mbed-os/pull/10692/commits/c115f23028c674329e859c26f4b7ad06f5d3d66c#diff-38f33b303ec6a94c6f64f1bd8778ca90R162

0xc0170 commented on features/netsocket/emac-drivers/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/network/whd_network.h:
```diff
@@ -0,0 +1,143 @@
+//typedef NOOS_DUMMY wiced_tls_context_t;
+//typedef NOOS_DUMMY wiced_tls_session_t;
+//typedef NOOS_DUMMY wiced_tls_certificate_t;
+//typedef NOOS_DUMMY  wiced_tls_endpoint_type_t;
```
dead code should be removed

vmedcy replied:
Removed whd_network.h, it was unused.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/cy_crc_api.c:
```diff
@@ -0,0 +1,91 @@
+
+void hal_crc_compute_partial_start(const crc_mbed_config_t *config)
+{
+    if (!cy_crc_initialized)
```
can you fix coding style in HAL implementation like this? see https://os.mbed.com/docs/mbed-os/v5.12/contributing/style.html

vmedcy replied:
Applied astyle (`astyle -n --options=.astylerc targets/TARGET_Cypress/TARGET_PSOC6/*.*`): https://github.com/ARMmbed/mbed-os/pull/10692/commits/b9f7836ac2bbb48acf4336ba1be5aa0da41cb990

Added style conformance check to CY internal CI pipeline.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/cy_serial_api.c:
```diff
@@ -0,0 +1,311 @@
+* You may use this file only in accordance with the license, terms, conditions, 
+* disclaimers, and limitations in the end user license agreement accompanying 
+* the software package with which this file was provided.
+********************************************************************************/
```
This header here should be fixed (the license - the previous files were apache, why not this one).

vmedcy replied:
Old license header was added by mistake, fixed in https://github.com/ARMmbed/mbed-os/pull/10692/commits/b9f7836ac2bbb48acf4336ba1be5aa0da41cb990#diff-0c19498117fa0dc69058c923815fb238R2

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/mbed_overrides.c:
```diff
@@ -53,3 +56,19 @@ void mbed_sdk_init(void)
+* Used here to initialize common parts of the Cypress libraries.
+*
+*******************************************************************************/
+void mbed_main(void)
```
shouldn't this be `mbed_sdk_init` ? mbed_main is users code hook

vmedcy replied:
Removed custom `mbed_main` from mbed_overrides.c.

Cannot execute `cybsp_wifi_init` in `mbed_sdk_init` since WHD initialization flow requires RTOS to be up and running.

morser499 commented on rtos/TARGET_CORTEX/mbed_rtx_conf.h:
```diff
@@ -59,7 +59,7 @@
 #endif
 
-#define OS_DYNAMIC_MEM_SIZE         0
+#define OS_DYNAMIC_MEM_SIZE        1024 
```
We know this change is wrong and are in the process of fixing the code that was dependent on this change

cypress-neil replied:
Removed in 
ec81b3e

cypress-neil commented on features/lwipstack/lwipopts.h:
```diff
@@ -83,9 +83,7 @@
-#ifndef LWIP_RAW
-#define LWIP_RAW                    0
-#endif
+#define LWIP_RAW                    1
```
These changes were added for PSOC6 testing but it is not clear if they should remain here as they affect all targets.
@vmedcy @morser499

0xc0170 replied:
It should not, this PR can be dependent on it (will be treated like that if required). I already mentioned few other issues like this, please create separate PR. They will get much faster in and easier to review.

vmedcy replied:
LWIP_RAW change moved to https://github.com/ARMmbed/mbed-os/pull/10978

cypress-neil commented on rtos/TARGET_CORTEX/mbed_rtx_conf.h:
```diff
@@ -59,7 +59,7 @@
 #endif
 
-#define OS_DYNAMIC_MEM_SIZE         0
+#define OS_DYNAMIC_MEM_SIZE        1024 
```
This was added for PSOC6 testing but it should be reverted once the PSOC6 code no longer requires it. @vmedcy @morser499

cypress-neil replied:
Removed in 
ec81b3e

0xc0170 commented on TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -71,7 +71,9 @@ void ticker_callback_1(void)
     ++callback_trigger_count;
-    switch_led2_state();
+    if (LED2 != NC) {
+        switch_led2_state();
```
Can you send this as separate PR - fixing this LED2 = NC condition.

vmedcy replied:
TESTS/mbed_drivers/ticker/main.cpp change is now submitted as https://github.com/ARMmbed/mbed-os/pull/10976

0xc0170 commented on TESTS/mbedmicro-rtos-mbed/malloc/main.cpp:
```diff
@@ -122,7 +122,7 @@ void test_alloc_and_free(void)
     int loop = ALLOC_LOOP;
     while (loop) {
-        data = malloc(size);
+        data = count < ALLOC_ARRAY_SIZE ? malloc(size) : NULL;
```
Also separate PR fixing this issue

vmedcy replied:
TESTS/mbedmicro-rtos-mbed/malloc/main.cpp is now submitted as https://github.com/ARMmbed/mbed-os/pull/10977

0xc0170 commented on features/lwipstack/LWIPStack.cpp:
```diff
@@ -309,14 +309,27 @@ nsapi_error_t LWIP::socket_open(nsapi_socket_t *handle, nsapi_protocol_t proto)
 #if LWIP_IPV6
     // Enable IPv6 (or dual-stack)
-    lwip_proto = (enum netconn_type)(lwip_proto | NETCONN_TYPE_IPV6);
+    netconntype = (enum netconn_type)(netconntype | NETCONN_TYPE_IPV6);
```
Another separate PR addressing LWIPStack

vmedcy replied:
features/lwipstack/LWIPStack.cpp change to add NSAPI_RAW support is now submitted as https://github.com/ARMmbed/mbed-os/pull/10978

0xc0170 commented on features/netsocket/RAWSocket.cpp:
```diff
@@ -0,0 +1,184 @@
+/*
+ * $ Copyright Cypress Semiconductor Apache2 $
```
fix the license header per our license guide

0xc0170 replied:
new feature like this needs to be in separate PR

vmedcy replied:
features/netsocket/RAWSocket.cpp addition is now tracked in https://github.com/ARMmbed/mbed-os/pull/10978

0xc0170 reviewed (changes requested):
Please separate fixes/new features. Send new PRs where they will be reviewed separately - included various teams as I read the changes here.

All these new PR can be referenced here and will be required for this PR (if they are), adding a list of dependencies helps us to track these (we set proceeding PR label here).

This PR should contain only target update. As it is, it is not `Target update` but the most significant is adding new functionality , thus would be `functionality change`.

vmedcy committed:
PSOC6: update Cypress targets to use psoc6csp

vmedcy committed:
PSOC6: update to latest psoc6pdl

vmedcy committed:
PSOC6: update BSP GeneratedSource, add design.modus

vmedcy committed:
PSOC6: update BSP sources

vmedcy committed:
PSOC6: add WHD sources

vmedcy committed:
PSOC6: add WHD EMAC driver

vmedcy committed:
PSOC6: update Cypress boards to use TARGET_WHD

vmedcy committed:
PSOC6: remove libwiced_drivers for targets ported to WHD

vmedcy committed:
TARGET_CYW43XXX: Added Bluetooth firmware

vmedcy committed:
PSOC6: remove PSA targets

vmedcy committed:
PSOC6: remove CY8CKIT-062S2-4343W (not ready for release)

Also removed _M0 targets that are no longer used.

0xc0170 commented on features/netsocket/emac-drivers/TARGET_WHD/interface/WhdSTAInterface.h:
```diff
@@ -0,0 +1,204 @@
+#ifndef WHD_STA_INTERFACE_H
+#define WHD_STA_INTERFACE_H
+
+#include "mbed.h"
```
include what is needed, not `mbed.h` file

vmedcy replied:
Replaced with:
```
#include "netsocket/WiFiInterface.h"
#include "netsocket/EMACInterface.h"
```

0xc0170 commented on features/netsocket/emac-drivers/TARGET_WHD/interface/WhdSTAInterface.h:
```diff
@@ -0,0 +1,204 @@
+     */
+    int set_olm(OlmInterface *olm) {
+        if (get_connection_status() == NSAPI_STATUS_DISCONNECTED)
+        {
```
`{` attached - run astyle on this file

vmedcy replied:
Re-applied astyle: 85a5bc0

0xc0170 commented on features/netsocket/emac-drivers/TARGET_WHD/interface/WhdSoftAPInterface.h:
```diff
@@ -0,0 +1,138 @@
+#ifndef WHD_SOFTAP_INTERFACE_H
+#define WHD_SOFTAP_INTERFACE_H
+
+#include "mbed.h"
```
remove

vmedcy replied:
Replaced with `#include "netsocket/EMACInterface.h"`

0xc0170 commented on features/netsocket/emac-drivers/TARGET_WHD/interface/whd_emac.cpp:
```diff
@@ -0,0 +1,272 @@
+{
+}
+
+WHD_EMAC &WHD_EMAC::get_instance() {
```
this file should follow Mbed OS style (please run astyle)

vmedcy replied:
Re-applied astyle: https://github.com/ARMmbed/mbed-os/pull/10692/commits/85a5bc0bf9eeac3565a933c46d1363825feb8d39

0xc0170 commented on usb/device/targets/TARGET_Cypress/TARGET_PSOC6/USBPhy_PSOC6.cpp:
```diff
@@ -0,0 +1,378 @@
+/*
+ * Copyright (c) 2019, Arm Limited and affiliates.
```
this PR still is adding additional functionaity (USB here) ? It should be also separate PR

vmedcy replied:
OK, removed commit that adds USB device implementation.

Is it appropriate to create pull request with USB device functionality BEFORE this PR is merged to mainline? Note, the implementation depends on psoc6hal and psoc6pdl additions introduced by this PR.

vmedcy committed:
PSOC6: apply astyle to WHD EMAC driver implementation

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/cy_port_api.c:
```diff
@@ -0,0 +1,75 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
```
these new files should be 2019 (wrong copy-paste): ` * Copyright (c) 2019, Arm Limited and affiliates.` would be perfect

vmedcy replied:
https://github.com/ARMmbed/mbed-os/pull/10692/commits/d77fb26e9f2d2f67fd3c6802ead4915e4dc11642: adjusted the copyright of new cy_*_api.c files, per the proposed template.

0xc0170 reviewed (changes requested):
One small cosmetic change requested, otherwise looks fine to me.

Functionality is the same as it was (tests all the same) ?

vmedcy committed:
PSOC6: fix license headers per review feedback

vmedcy committed:
Merge remote-tracking branch 'ARMmbed/master' into pr/psoc6-hal

AnttiKauppila commented on targets/TARGET_Cypress/TARGET_PSOC6/PeripheralPins.h:
```diff
@@ -68,4 +68,13 @@ extern const PinMap PinMap_ADC[];
 extern const PinMap PinMap_DAC[];
 #endif
 
+#if DEVICE_QSPI
```
This should be combined with above (line 55-59)

vmedcy replied:
This is addressed by https://github.com/ARMmbed/mbed-os/pull/10692/commits/bab34cb4675ced134deaae178440c20fb9cd9779

Note: DEVICE_QSPI is not yet enabled for Cypress targets (cy_qspi_api.c will not compile when device_has: "QSPI" added to targets.json). This will be addressed in subsequent PR.

AnttiKauppila replied:
ok

vmedcy committed:
Merge remote-tracking branch 'ARMmbed/master' into pr/psoc6-hal

vmedcy committed:
PSOC6: cleanup DEVICE_QSPI mappings

Note: device_has: "QSPI" is still disabled for TARGET_PSOC6
(QSPI HAL implementation is incomplete).

vmedcy committed:
PSOC6: do not remove QSPI from FUTURE_SEQUANA targets

Since QSPi is not yet supported by base TARGET_PSOC6,
there is no need to remove the device label from
FUTURE targets that inherit from TARGET_PSOC6.

This will need to be reverted back once the QSPI support
is implemented for Cypress PSOC6 targets.

AnttiKauppila commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/resources/firmware/TARGET_43012/resources.h:
```diff
@@ -0,0 +1,28 @@
+#include "wiced_resource.h"
+
+extern const resource_hnd_t wifi_firmware_image;
+extern const char wifi_firmware_image_data[360823];
```
This look like a really big value, is this intentional. What is this magic number anyway?

vmedcy replied:
@AnttiKauppila: this is generated file, as well as the firmware image itself - [43012C0_bin.c](https://github.com/ARMmbed/mbed-os/pull/10692/files/d77fb26e9f2d2f67fd3c6802ead4915e4dc11642#diff-1ef39d778285ff9127a03db0d408ae5f) - it generated from the binary blob that is loaded to the CYW43012 WiFi Radio: [43012C0.bin](https://github.com/ARMmbed/mbed-os/pull/10692/files/d77fb26e9f2d2f67fd3c6802ead4915e4dc11642#diff-9c4b8b71db880f2b410b723ec317768a).

AnttiKauppila commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/src/bus_protocols/whd_bus_sdio_protocol.c:
```diff
@@ -0,0 +1,1427 @@
+* Enable this section for logging of SDIO transfers
+* by changing "if 0" to "if 1"
+******************************************************/
+#if 0
```
Maybe this should be put behind a configuration flag

0xc0170 replied:
+1 but looking at the file path, this might is in their driver (3rd party)

vmedcy replied:
This code is used internally by WHD driver developers to debug SDIO transfers. There is no value in exposing this as configuration flag for customers.

AnttiKauppila commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_WHD/src/bus_protocols/whd_bus_sdio_protocol.c:
```diff
@@ -0,0 +1,1427 @@
+        CHECK_RETURN(whd_chip_specific_socsram_init(whd_driver) );
+    }
+
+#if 0
```
Is this code needed anymore? Could it be put behind a flag?

vmedcy replied:
This code is used internally to debug SDIO bus transfers. It is not related to Mbed.

vmedcy committed:
PSOC6: TARGET_PSOC6_FUTURE doesn't implement DEVICE_I2CSLAVE

AnttiKauppila reviewed (approved):
Went through the code and left couple of questions. Those won't be blocking this PR.
